crates/tabby/src/services/chat.rs

```diff
@@ -75,13 +76,26 @@ impl ChatCompletionChunk {
 
 pub struct ChatService {
     engine: Arc<dyn TextGeneration>,
+    pub(crate) logger: Arc<dyn EventLogger>,
     prompt_builder: ChatPromptBuilder,
 }
 
+impl ChatCompletionRequest {
+    pub fn prompt(&self) -> String {
+        // Doubt this is the right way to find the prompt.
+        self.messages.iter().map(|m| &*m.content).collect()
+    }
+}
+
 impl ChatService {
-    fn new(engine: Arc<dyn TextGeneration>, chat_template: String) -> Self {
+    fn new(
+        engine: Arc<dyn TextGeneration>,
+        logger: Arc<dyn EventLogger>,
+        chat_template: String,
+    ) -> Self {
         Self {
             engine,
+            logger,
             prompt_builder: ChatPromptBuilder::new(chat_template),
         }
     }
```

Comment:
i recommend implement the logging logic inside of `generate` function here, so you don’t have to expose logger as public.

This is also what we done for `completion.rs` service

Reply:
Looking into this and finding that the issue is we'll have to serialize each chunk twice. Once in the generator to create the log output lines, and once in the endpoint to create the response. Is this an acceptable overhead?

Reply:
For now I will just push the string content into a Vec.